app/InputHandler.js

```diff
@@ -149,11 +149,31 @@ $.extend( InputHandler.prototype, {
 			forcePrefix = true;
 		}
 
-		var segments = url.split( '/' );
+		var filename;
 
-		var filename = ( $.inArray( 'thumb', segments ) !== -1 )
-			? segments[segments.length - 2]
-			: segments[segments.length - 1];
+		var mediaKey = '#mediaviewer/';
+		var mediaKeyLoc = url.indexOf( mediaKey );
+
+		var fileKey = 'File:';
+		var fileKeyLoc = url.indexOf( fileKey );
+
+		if( mediaKeyLoc !== -1 ) {
+			filename = url.substr( mediaKeyLoc + mediaKey.length );
+		} else if( fileKeyLoc !== -1 ) {
+			filename = url.substr( fileKeyLoc );
+		} else {
```

Comment:
I suspect you can nicely split up this method. For instance the body of this else looks like can be extracted into something cohesive.